Migrate processor/feature detection from OpenJ9 to OMR Part 1

AidanHa opened the pull request:
OpenJ9 port library has processor and feature detections that can be sunk down to OMR. This can be divided into 4 steps:
1: Creation of processor and feature detection in OMR (from OpenJ9).
2: Implementation of the new feature throughout out OMR.
3: Implementation of the new feature throughout OpenJ9.
4: Deletion of the processor detection in OpenJ9 port library.

This PR serves to implement step 1 of the list above. This PR includes the following:

- Move and rename necessary port library macros from OpenJ9 to OMR's `omrport.h` (Commit 1)
- Move and implement processor/feature detection for windows systems[WIP] (Commit 2)
- Move and implement processor/feature detection for unix systems (x86, Power, AIX, Z) (Commit 3)
- Move and test the respective port tests that helps verify that processor detection works (Commit 4)

Issue: #4339

AidanHa commented:
These functions have been tested on Z, Power and X86 for Windows and Unix systems! I think we're ready for a full review!

Leonardo2718 commented:
@genie-omr build all

AidanHa commented:
@Leonardo2718 Could you run the PR builds again? There were some build issues on LinuxPPC and AIX, and I've pushed changes to fix these issues.

AidanHa commented:
Also FYI, I've broken down the PR into several commits for clarity, but will merge them when the changes are approved!

rwy0717 commented:
@genie-omr build all

rwy0717 commented:
Aidan, for the new tests that you're adding, you should just use core gtest functionality, rather than using the port test helpers. Specifically, don't call `reportTestEntry` (or exit), and use `ASSERT`/`EXPECT`, rather than calling `outputErrorMessage`. As well, your tests should be in `TEST` blocks (I'm looking at `testGetProcessorDescription`).

rwy0717 commented:
This looks like a pretty straightforward copy/paste/rename job, so if you want to do clean up in a later PR, I would be fine with that, it's up to you.

rwy0717 commented on port/common/omrsysinfo.c:
```diff
@@ -89,6 +89,35 @@ omrsysinfo_get_CPU_architecture(struct OMRPortLibrary *portLibrary)
+intptr_t
+omrsysinfo_get_processor_description(struct OMRPortLibrary *portLibrary, OMRProcessorDesc *desc)
+{
+	return -1;
```
This should be OMRPORT_ERROR_NOT_SUPPORTED_ON_THIS_PLATFORM

rwy0717 commented on port/common/omrsysinfo_helpers.c:
```diff
@@ -0,0 +1,213 @@
+#define CPUID_FAMILYCODE_AMDATHLON		0x06
+#define CPUID_FAMILYCODE_AMDOPTERON		0x0F
+
+#define CPUID_MODELCODE_AMDK5			0x04
```
When aligning, use spaces, not tabs.

rwy0717 commented on port/win32/omrsysinfo.c:
```diff
@@ -90,6 +91,40 @@ omrsysinfo_get_CPU_architecture(struct OMRPortLibrary *portLibrary)
+
+	if ((NULL != desc)
+	&& (feature < (OMRPORT_SYSINFO_OS_FEATURES_SIZE * 32))
+	) {
```
This formatting is... interesting :-)

AidanHa replied:
Sorry, not too sure what happened here

AidanHa replied:
Should we also do this for `omrsysinfo_get_processor_description`

rwy0717 commented on port/common/omrsysinfo.c:
```diff
@@ -89,6 +89,35 @@ omrsysinfo_get_CPU_architecture(struct OMRPortLibrary *portLibrary)
+omrsysinfo_processor_has_feature(struct OMRPortLibrary *portLibrary, OMRProcessorDesc *desc, uint32_t feature)
+{
+	return FALSE;
+}
```
It may be a little silly, but you should still be invoking the enter/exit tracepoints in this fallback implementation.

AidanHa replied:
Should we also do this for `omrsysinfo_get_processor_description`

rwy0717 replied:
Yep.

rwy0717 commented on port/zos390/omrportpg.h:
```diff
@@ -48,6 +48,17 @@
+	uintptr_t lastDoubleWord;
+	STFLEFacilities facilities;
+} OMRSTFLECache;
+
```
Should these structs be under an OMR_ENV_DATA64 ifdef?

AidanHa replied:
There was no ifdef in the openJ9 equivalent, so I assumed it wasn't necessary... Should I add one?

rwy0717 replied:
It seems to me that these structures are only used in 64bit ZOS, so I am thinking yes.

rwy0717 commented on port/zos390/omrportpg.h:
```diff
@@ -93,4 +105,6 @@ typedef struct OMRPortPlatformGlobals {
 #define PPG_si_executableName (portLibrary->portGlobals->platformGlobals.si_executableName)
 #define PPG_ipt_ttoken (portLibrary->portGlobals->platformGlobals.iptTtoken)
 
+#define PPG_stfleCache (portLibrary->portGlobals->platformGlobals.stfleCache)
```
Should this macro be wrapped under an OMR_ENV_DATA64 ifdef?

AidanHa replied:
Similarly there was no ifdef in the openJ9 equivalent

rwy0717 commented:
@genie-omr build aix

AidanHa commented:
> Aidan, for the new tests that you're adding, you should just use core gtest functionality, rather than using the port test helpers. Specifically, don't call `reportTestEntry` (or exit), and use `ASSERT`/`EXPECT`, rather than calling `outputErrorMessage`. As well, your tests should be in `TEST` blocks (I'm looking at `testGetProcessorDescription`).

I removed `testGetProcessorDescription` and added the tests in a `TEST` block. Any reason why we want to be using `ASSERTS` rather than `outputErrorMessage`? When writing the tests, I was following the format of the other tests on `si.cpp`

rwy0717 commented:
Wherever possible, the port test helpers are being replaced with gtest. I opened an issue https://github.com/eclipse/omr/issues/4530. Any new tests should just be using gtest.

rwy0717 commented on port/zos390/omrportpg.h:
```diff
@@ -48,6 +48,17 @@
+	uint64_t dw1;
+	uint64_t dw2;
+	uint64_t dw3;
+} STFLEFacilities;
```
STFLEFacilities should start with OMR.

rwy0717 commented on port/unix/omrsysinfo.c:
```diff
@@ -531,6 +576,816 @@ omrsysinfo_get_CPU_architecture(struct OMRPortLibrary *portLibrary)
+		setFeature(desc, OMRPORT_PPC_FEATURE_POWER4);
+	}
+#endif /* !defined(J9OS_I5_V6R1) */
+#if !defined(J9OS_I5_V7R2) && !defined(J9OS_I5_V6R1)
```
These are never defined in OMR, and the code below fails to compile on aix 6.1.

rwy0717 commented on port/common/omrsysinfo.c:
```diff
@@ -89,6 +89,35 @@ omrsysinfo_get_CPU_architecture(struct OMRPortLibrary *portLibrary)
 	return "unknown";
 #endif
 }
+/**
```
Missing a new line before this comment

rwy0717 commented:
@genie-omr build all

fjeremic commented on compiler/arm/codegen/ControlFlowEvaluator.cpp:
```diff
@@ -1812,6 +1812,7 @@ static TR::Register *generateMaxMin(TR::Node *node, TR::CodeGenerator *cg, bool
 
    TR_ARMConditionCode cond = max ? (isUnsigned ? ARMConditionCodeHI : ARMConditionCodeGT) : (isUnsigned ? ARMConditionCodeLS : ARMConditionCodeLT);
    int n = node->getNumChildren();
+   TR_ASSERT(n == 2, "Expected 2 children only");
```
Is this change related to the PR? Similarly for the other `TR_ASSERT` changes?

rwy0717 replied:
I think Aidan has accidentally pushed a few extra commits, these compiler changes weren't present initially.

AidanHa replied:
Apologies, I did a rebase off of my master, and that branch had some extra commits

fjeremic commented on port/common/omrsysinfo_helpers.h:
```diff
@@ -0,0 +1,39 @@
+getX86CPUID(uint32_t leaf, uint32_t *cpuInfo);
+
+extern intptr_t
+getX86Description(struct OMRPortLibrary *portLibrary, OMRProcessorDesc *desc);
```
The corresponding Z `getstfle` function which performs similar operations is in the `port/linuxs390` directory. Should this go into `port/linux386`? It does not look like common code, so to me it is odd that it is currently in `port/common`.

AidanHa replied:
I just copied the implementation of OpenJ9's port library,  but I do believe `getX86Description` is used no only in both `win32/omrsysinfo.c` as well as in `unix/omrsysinfo.c`. Not too sure why it was done like that in OpenJ9, but I'm assuming that we might get a compilation error if we add `getX86Description` to `port/linux386` (even if the call is protected by an ifdef...)

rwy0717 replied:
Yes, because the port library is organized primarily by OS, code common to multiple OSes must be placed in common.

fjeremic commented:
@AidanHa may I suggest adding new commits addressing reviews? It is quite difficult given the size of the change for reviewers to determine whether reviews were addressed without having to analyze the entire change once something is force pushed.

@rwy0717 can we start resolving the review comments if they have been addressed so we know what remains to be fixed in the PR?

rwy0717 commented on port/common/omrsysinfo_helpers.c:
```diff
@@ -0,0 +1,213 @@
+ *							4 characters of the the Hypervisor Vendor String
+ *              cpuInfo[3]	For EAX = 0x40000000 hold the last 4 characters of the
+ *              			Hypervisor Vendor String
+ *
```
Use spaces to align tables of text.

rwy0717 commented on port/common/omrport.tdf:
```diff
@@ -1093,3 +1093,9 @@ TraceException=Trc_PRT_sysinfo_get_open_file_count_memAllocFailed Group=sysinfo
+TraceExit=Trc_PRT_sysinfo_get_processor_description_Exit Group=j9sysinfo Overhead=1 Level=5 NoEnv Template="sysinfo_get_processor_description: returning with %zd"
+
+TraceEntry=Trc_PRT_sysinfo_processor_has_feature_Entered Group=j9sysinfo Overhead=1 Level=5 NoEnv Template="sysinfo_processor_has_feature: desc = %p, feature = %d"
+TraceExit=Trc_PRT_sysinfo_processor_has_feature_Exit Group=j9sysinfo Overhead=1 Level=5 NoEnv Template="sysinfo_processor_has_feature: returning with %zu."
```
I believe the group should just be "sysinfo", sans j9.

rwy0717 commented on fvtest/porttest/si.cpp:
```diff
@@ -2343,3 +2343,40 @@ TEST(PortSysinfoTest, sysinfo_cgroup_get_memlimit)
+	
+	for (i = 0; i < OMRPORT_SYSINFO_FEATURES_SIZE * 32; i++) {
+		BOOLEAN feature = omrsysinfo_processor_has_feature(&desc, i);
+		ASSERT_FALSE((TRUE != feature) && (FALSE != feature));
```
This seems like a lot of negation when something like the below would suffice.
```
EXPECT_TRUE(feature == TRUE || feature == FALSE);
```

rwy0717 commented on fvtest/porttest/si.cpp:
```diff
@@ -2343,3 +2343,40 @@ TEST(PortSysinfoTest, sysinfo_cgroup_get_memlimit)
+	int32_t rc = 0;
+	int i = 0;
+	
+	reportTestEntry(OMRPORTLIB, testName);
```
Don't report test entry or exit.

rwy0717 commented on port/zos390/omrportpg.h:
```diff
@@ -46,7 +46,20 @@
 #define TTOKEN_BUF_SZ                   16
-#endif /* defined(OMR_ENV_DATA64) */
+
+
```
Extra newline here, and below at line 56.

rwy0717 commented on port/linuxs390/omrsysinfo_helpers.h:
```diff
@@ -0,0 +1,43 @@
+/*******************************************************************************
+ * Copyright (c) 2013, 2013 IBM Corp. and others
```
Copyright is wrong, should be 2013, 2019

rwy0717 commented on fvtest/porttest/si.cpp:
```diff
@@ -2343,3 +2343,37 @@ TEST(PortSysinfoTest, sysinfo_cgroup_get_memlimit)
+	ASSERT_FALSE(rc == -1);
+
+#if (defined(J9X86) || defined(J9HAMMER))
+	ASSERT_FALSE(desc.processor < PROCESSOR_X86_UNKNOWN);
```
This test could be:
```C++
ASSERT_GE(desc.processor, PROCESSOR_X86_UNKNOWN);
```
This way, In the case of a test failure, gtest will be able to print what the two compared values are.

rwy0717 commented on fvtest/porttest/si.cpp:
```diff
@@ -2343,3 +2343,37 @@ TEST(PortSysinfoTest, sysinfo_cgroup_get_memlimit)
+	ASSERT_FALSE(desc.processor < PROCESSOR_X86_UNKNOWN);
+	ASSERT_FALSE(desc.physicalProcessor < PROCESSOR_X86_UNKNOWN);
+#elif (defined(AIXPPC) || defined(LINUXPPC))
+	ASSERT_FALSE((desc.processor < PROCESSOR_PPC_UNKNOWN) || (desc.processor >= PROCESSOR_X86_UNKNOWN));
```
This check can be split into two much simpler conditions:
```c++
ASSERT_GE(desc.processor, PROCESSOR_PPC_UNKNOWN);
ASSERT_LT(desc.processor, PROCESSOR_X86_UNKNOWN);
```

rwy0717 commented on fvtest/porttest/si.cpp:
```diff
@@ -2343,3 +2343,37 @@ TEST(PortSysinfoTest, sysinfo_cgroup_get_memlimit)
+	rc = omrsysinfo_get_processor_description(&desc);
+	ASSERT_FALSE(rc == -1);
+
+#if (defined(J9X86) || defined(J9HAMMER))
```
The outer parens are superfluous. The `#elif`/`#else` below need to be fixed up as well.

rwy0717 commented on fvtest/porttest/si.cpp:
```diff
@@ -2343,3 +2343,37 @@ TEST(PortSysinfoTest, sysinfo_cgroup_get_memlimit)
+		ASSERT_TRUE(feature == TRUE || feature == FALSE);
+	}
+
+	return;
```
No need to explicitly return from a void function.

rwy0717 commented on fvtest/porttest/si.cpp:
```diff
@@ -2343,3 +2343,37 @@ TEST(PortSysinfoTest, sysinfo_cgroup_get_memlimit)
+TEST(PortSysinfoTest, sysinfo_test_sysinfo_get_processor_description)
+{
+	OMRPORT_ACCESS_FROM_OMRPORT(portTestEnv->getPortLibrary());
+	const char *testName = "omrsysinfo_get_processor_description";
```
testName is unused.

rwy0717 commented on fvtest/porttest/si.cpp:
```diff
@@ -2343,3 +2343,37 @@ TEST(PortSysinfoTest, sysinfo_cgroup_get_memlimit)
+	OMRPORT_ACCESS_FROM_OMRPORT(portTestEnv->getPortLibrary());
+	const char *testName = "omrsysinfo_get_processor_description";
+	int32_t rc = 0;
+	int i = 0;
```
This is C++, don't declare locals up front. This `i` local can be declared in the for loop below. The `rc` should be beside the declaration for `desc`

rwy0717 commented on fvtest/porttest/si.cpp:
```diff
@@ -2343,3 +2343,37 @@ TEST(PortSysinfoTest, sysinfo_cgroup_get_memlimit)
+/**
+ * Test omrsysinfo_get_processor_description.
+ */
+TEST(PortSysinfoTest, sysinfo_test_sysinfo_get_processor_description)
```
Unfortunately, gtest style guidelines say you should not use underscores in test names (it's in the [FAQ](https://github.com/google/googletest/blob/master/googletest/docs/faq.md#why-should-test-suite-names-and-test-names-not-contain-underscore)). Further, this test name is super verbose. A better test name would be:
```c++
TEST(PortSysinfoTest, GetProcessorDescription)
```

rwy0717 commented on port/unix/omrsysinfo.c:
```diff
@@ -531,6 +576,810 @@ omrsysinfo_get_CPU_architecture(struct OMRPortLibrary *portLibrary)
+{
+	if ((NULL != desc)
+	&& (feature < (OMRPORT_SYSINFO_FEATURES_SIZE * 32))
+	) {
```
Bad formatting.

rwy0717 commented on port/unix/omrsysinfo.c:
```diff
@@ -531,6 +576,810 @@ omrsysinfo_get_CPU_architecture(struct OMRPortLibrary *portLibrary)
+
+	/* initialize auxv prior to querying the auxv */
+	if (prefetch_auxv()) {
+_error:
```
Place the error block at the end of the function, and goto _error from here. This way, there are no backwards gotos.

AidanHa replied:
How would this work? We swould still need to call `prefetch_auxv()` at the beginning of the function... Perhaps i replace  the goto_error with the error block and vice versa?

AidanHa replied:
`/* initialize auxv prior to querying the auxv */
	if (prefetch_auxv()) {
                goto _error;
	}

	/* Linux PPC processor */
	platform = (char *) query_auxv(AT_PLATFORM);
	if ((NULL == platform) || (((char *) -1) == platform)) {
_error:
		desc->processor = OMR_PROCESSOR_PPC_UNKNOWN;
		desc->physicalProcessor = OMR_PROCESSOR_PPC_UNKNOWN;
		desc->features[0] = 0;
		desc->features[1] = 0;
		return -1;
	}`

rwy0717 replied:
You would put the error handling all the way after the successful return:
```c++
at the end of the function...
	return 0;

on_error:
  desc->processor = OMR_PROCESSOR_PPC_UNKNOWN;
  desc->physicalProcessor = OMR_PROCESSOR_PPC_UNKNOWN;
  desc->features[0] = 0;
  desc->features[1] = 0;
  return -1;
}
```

rwy0717 commented on port/unix/omrsysinfo.c:
```diff
@@ -531,6 +576,810 @@ omrsysinfo_get_CPU_architecture(struct OMRPortLibrary *portLibrary)
+	char* base_platform = NULL;
+
+	/* initialize auxv prior to querying the auxv */
+	if (prefetch_auxv()) {
```
For clarity, this should be `!= 0`, because this function returns an error code, not a boolean.

rwy0717 commented on fvtest/porttest/si.cpp:
```diff
@@ -2343,3 +2343,32 @@ TEST(PortSysinfoTest, sysinfo_cgroup_get_memlimit)
+	ASSERT_LT(desc.physicalProcessor, OMR_PROCESSOR_X86_UNKNOWN);
+#elif defined(S390) || defined(J9ZOS390)
+	ASSERT_FALSE(desc.processor >= (OMR_PROCESSOR_PPC_UNKNOWN));
+	ASSERT_FALSE(desc.physicalProcessor >= (OMR_PROCESSOR_PPC_UNKNOWN));
```
These assertions can also be fixed up to use `ASSERT_LT/GT`.

AidanHa replied:
Whoops, must've missed these

rwy0717 commented on port/common/omrsysinfo_helpers.c:
```diff
@@ -0,0 +1,211 @@
+#define CPUID_FAMILYCODE_AMDOPTERON         0x0F
+
+#define CPUID_MODELCODE_AMDK5               0x04
+/**
```
Missing newline.

rwy0717 commented on port/win32/omrsysinfo.c:
```diff
@@ -90,6 +91,39 @@ omrsysinfo_get_CPU_architecture(struct OMRPortLibrary *portLibrary)
+	Trc_PRT_sysinfo_processor_has_feature_Entered(desc, feature);
+
+	if ((NULL != desc) && (feature < (OMRPORT_SYSINFO_OS_FEATURES_SIZE * 32)))
+	{
```
Formatting is still off, brace should be on the same line.

AidanHa replied:
So like `if ((NULL != desc) && (feature < (OMRPORT_SYSINFO_OS_FEATURES_SIZE * 32))) {`? Why is that?

rwy0717 replied:
> So like if ((NULL != desc) && (feature < (OMRPORT_SYSINFO_OS_FEATURES_SIZE * 32))) {?

Yep :)

> Why is that?

Dunno, but it's in the [guide](https://github.com/eclipse/omr/blob/master/doc/CodingStandard.md#brace-style).

rwy0717 commented on port/unix/omrsysinfo.c:
```diff
@@ -531,6 +576,810 @@ omrsysinfo_get_CPU_architecture(struct OMRPortLibrary *portLibrary)
+
+#endif /* defined(AIXPPC) */
+
+#if (defined(S390) || defined(J9ZOS390))
```
I'm not sure but, is this if missing `|| defined(J9ZTPF))` ?

AidanHa replied:
There was no `|| defined(J9ZTPF))` in the J9 version, so I left it as is

rwy0717 commented on port/unix/omrsysinfo.c:
```diff
@@ -531,6 +576,810 @@ omrsysinfo_get_CPU_architecture(struct OMRPortLibrary *portLibrary)
+
+	if (testSTFLE(portLibrary, 0)) {
+#if (defined(S390) && defined(LINUX))
+	/* OS Support for Linux on Z */
```
Indentation is wrong, but I'm not convinced this comment is even needed. There are a few similar comments below, with the same issue.

AidanHa replied:
I fixed all instances of this

rwy0717 commented on port/unix/omrsysinfo.c:
```diff
@@ -531,6 +576,810 @@ omrsysinfo_get_CPU_architecture(struct OMRPortLibrary *portLibrary)
+
+	if ((NULL != desc)
+	&& (feature < (OMRPORT_SYSINFO_FEATURES_SIZE * 32))
+	) {
```
Bad formatting.

rwy0717 commented on port/unix/omrsysinfo.c:
```diff
@@ -187,6 +195,43 @@ static uintptr_t copyEnvToBufferSignalHandler(struct OMRPortLibrary *portLib, ui
 
+#if (defined(LINUXPPC) || defined(AIXPPC))
+static OMRProcessorArchitecture mapPPCProcessor(const char *processorName);
+
```
This blank line is out of place.

rwy0717 commented on port/unix/omrsysinfo.c:
```diff
@@ -71,22 +72,29 @@
 #endif /* defined(J9ZOS390) */
 
-#if defined(LINUXPPC)
+#if (defined(LINUXPPC) || (defined(S390) && defined(LINUX) && !defined(J9ZTPF)))
```
outermost parens not needed.

rwy0717 commented on port/common/omrsysinfo_helpers.h:
```diff
@@ -0,0 +1,39 @@
+extern intptr_t
+getX86Description(struct OMRPortLibrary *portLibrary, OMRProcessorDesc *desc);
+
+#endif /* SYSINFOHELPERS_H_ */
```
Missing newline.

AidanHa commented:
Changes are fixed! @rwy0717 When you are done reviewing, I'll merge all of my changes into 1 single commit.

rwy0717 commented:
@genie-omr build all

rwy0717 commented:
@genie-omr build all

fjeremic commented:
Let's make sure to squash the review addressing commits here before merging.

rwy0717 commented:
@genie-omr build all

rwy0717 commented:
@fjeremic when you say "Let's make sure to squash", are you asking @AidanHa to squash, or are you asking me to use the "squash and merge" button? Either way, I think this PR is ready to go.

AidanHa commented:
I've gone ahead and merged my commits!

rwy0717 commented:
@genie-omr build all
